pkg/diagnostics/types/diagnostic.go

```diff
@@ -0,0 +1,175 @@
+package types
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"runtime"
+	"strings"
+
+	"github.com/openshift/origin/pkg/diagnostics/log"
+)
+
+type Diagnostic interface {
+	Name() string
+	Description() string
+	CanRun() (canRun bool, reason error)
+	Check() *DiagnosticResult
+}
+
+type DiagnosticResult struct {
+	failure  bool
+	origin   string // name of diagnostic; automatically inserted into log Entries
+	logs     []log.Entry
+	warnings []DiagnosticError
+	errors   []DiagnosticError
+}
+
+func NewDiagnosticResult(origin string) *DiagnosticResult {
+	return &DiagnosticResult{origin: origin}
+}
+
+func (r *DiagnosticResult) Complete() *DiagnosticResult {
+	if r.errors == nil {
+		r.errors = make([]DiagnosticError, 0)
+	}
+	if r.warnings == nil {
+		r.warnings = make([]DiagnosticError, 0)
+	}
+	if r.logs == nil {
+		r.logs = make([]log.Entry, 0)
+	}
+	return r
+}
+
+func (r *DiagnosticResult) appendLogs(stackDepth int, entry ...log.Entry) {
+	if r.logs == nil {
+		r.logs = make([]log.Entry, 0)
+	}
+	r.logs = append(r.logs, entry...)
+	// glog immediately for debugging when a diagnostic silently chokes
+	for _, entry := range entry {
+		if glog.V(glog.Level(6 - entry.Level.Level)) {
+			glog.InfoDepth(stackDepth, entry.Message.String())
+		}
+	}
+}
+
+func (r *DiagnosticResult) Failure() bool {
+	return r.failure
+}
+
+func (r *DiagnosticResult) Logs() []log.Entry {
+	if r.logs == nil {
+		return make([]log.Entry, 0)
+	}
+	return r.logs
+}
+
+func (r *DiagnosticResult) appendWarnings(warn ...DiagnosticError) {
+	if r.warnings == nil {
+		r.warnings = make([]DiagnosticError, 0)
+	}
+	r.warnings = append(r.warnings, warn...)
+}
+
+func (r *DiagnosticResult) Warnings() []DiagnosticError {
+	if r.warnings == nil {
+		return make([]DiagnosticError, 0)
+	}
+	return r.warnings
+}
+
+func (r *DiagnosticResult) appendErrors(err ...DiagnosticError) {
+	if r.errors == nil {
+		r.errors = make([]DiagnosticError, 0)
+	}
+	r.failure = true
+	r.errors = append(r.errors, err...)
+}
+
+func (r *DiagnosticResult) Errors() []DiagnosticError {
+	if r.errors == nil {
+		return make([]DiagnosticError, 0)
+	}
+	return r.errors
+}
+
+func (r *DiagnosticResult) Append(r2 *DiagnosticResult) {
```

Comment:
You're right. And the one place actually using this should probably do it a little differently.